dacbench/envs/layerwise_sgd.py

```diff
@@ -0,0 +1,539 @@
+            test_accuracies.append(torch.sum(preds == d_target) / len(d_target))
+            i += 1
+            if i >= nmb_sets:
+                break
+    return (
+        torch.cat(test_losses).cpu().numpy(),
+        torch.tensor(test_accuracies).cpu().numpy(),
+    )
+
+
+class LayerwiseSGDEnv(AbstractMADACEnv):
+    """The SGD DAC Environment implements the problem of dynamically configuring
+    the learning rate hyperparameter of a neural network optimizer
+    (more specifically, torch.optim.AdamW) for a supervised learning task.
+    While training, the model is evaluated after every epoch.
+
+    Actions correspond to learning rate values in [0,+inf[
+    For observation space check `observation_space` method docstring.
+    For instance space check the `SGDInstance` class docstring
+    Reward:
+        negative loss of model on test_loader of the instance       if done
+        crash_penalty of the instance                               if crashed
+        0                                                           otherwise
+    """
+
+    metadata = {"render_modes": ["human"]}  # noqa: RUF012
+
+    def __init__(self, config):
+        """Init env."""
+        super().__init__(config)
+        torch.use_deterministic_algorithms(True) # For reproducibility
+        self.epoch_mode = config.get("epoch_mode", True)
+        self.device = config.get("device")
+
+        self.optimizer_type = torch.optim.SGD
+        self.optimizer_params = config.get("optimizer_params")
+        self.batch_size = config.get("training_batch_size")
+        self.model = config.get("model")
+        self.crash_penalty = config.get("crash_penalty")
+        self.loss_function = config["loss_function"](**config["loss_function_kwargs"])
+        self.dataset_name = config.get("dataset_name")
+        self.use_generator = config.get("model_from_dataset")
+        self.torchub_model = config.get("torch_hub_model", (False, None, False))
+
+        # Use default reward function, if no specific function is given
+        self.get_reward = config.get("reward_function", self.get_default_reward)
+
+        # Use default state function, if no specific function is given
+        self.get_states = config.get("state_method", self.get_default_states)
+
+        self.initial_learning_rate = config.get("initial_learning_rate")
+        self.state_version = config.get("state_version")
+        self.initial_seed = config.get("seed")
+        self.seed(self.initial_seed)
+
+        self.instance_set_path = config.get("instance_set_path")
+        self.fraction_of_dataset = config.get("fraction_of_dataset")
+        self.train_validation_ratio = config.get("train_validation_ratio")
+        self.instance_mode = config.get("instance_mode")
+        self.instance_set = config.get("instance_set")
+        self.inst_id = 0
+
+        self.predictions = deque(torch.zeros(2))
+
+    def step(self, action: list[float]):
```

Comment:
out of consistency, we should name them actions (or would Mrs Liskov be angry with us?

Reply:
She would, probably. But honestly, idc :D